Simplify code to remove named markdown section

linearcombination opened the pull request:
Legacy method that removed specified markdown sections was highly
imperative and exhibits the boolean flag anti-pattern. Replaced with
simple declarative regex expression.

linearcombination committed:
Simplify code to remove named markdown section

Legacy method that removed specified markdown sections was highly
imperative and exhibits the boolean flag anti-pattern. Replaced with
simple declarative regex expression.

PurpleGuitar commented on backend/document/markdown_extensions/remove_section_preprocessor.py:
```diff
@@ -27,28 +27,17 @@ def remove_sections(self, md: str) -> list[str]:
+        md: str,
+        section_name: str,
+        # See https://regex101.com/r/SSItAG/1
+        section_regex_fmt_str: str = r"\#+ +{}.*$\n*.*",
```
This regex seems like it would match the section header and the rest of the document, not just the rest of the section.  Does this have any tests that demonstrate correct behavior?

linearcombination replied:
Good catch, thank you!

linearcombination replied:
I updated this PR with quite a few changes. Here is the commit message for the latest push to this PR:

"Refactor remove_section_preprocessor to follow the pattern that the
other markdown preprocessor plugins follow so that the parameters the
plugin needs are passed in at construction time for the sake of
referential transparency.

Existing test, https://github.com/linearcombination/DOC/blob/1e803185f4f152aa40b3f10e5bd638f7b69b4a64/tests/unit/test_markdown_extensions.py#L28,  was improved to ensure that markdown sections which are
not to be removed and occur after sections that are to be removed
remain intact after processing by this preprocessor.

Regex approach used in remove_md_section was abandoned in favor of
legacy algo since regex complexity required to yield desired result is
arguably more complex to grok for newcomers. Edit: The regex you commented on did indeed extend past the markdown sections to remove. 
"

linearcombination committed:
Refactor remove_section_preprocessor module and update existing test

Refactor remove_section_preprocessor to follow the pattern that the
other markdown preprocessor plugins follow so that the parameters the
plugin needs are passed in at construction time for the sake of
referential transparency.

Existing test was improved to ensure that markdown sections which are
not to be removed and occur after sections that are to be removed
remain intact after processing by this preprocessor.

Regex approach used in remove_md_section was abandoned in favor of
legacy algo since regex complexity required to yield desired result is
arguably more complex to grok for newcomers.

linearcombination committed:
Add prefixes for additional markdown sections to be removed

So called 'Picture of' markdown sections in non-usfm files
often (always?) contain invalid (as in 404) URLs for pictures that
could otherwise be inserted into the document. So for now at least,
removing such markdown sections. The test referenced in this PR covers
their testing.

PurpleGuitar reviewed (approved):
Looks good.  Nice work!
